File: backend/app/models/task.py
import enum
from datetime import datetime

from sqlalchemy import Boolean, DateTime, Enum, Float, Integer, String, Text, UniqueConstraint, func
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class TaskStatus(str, enum.Enum):
    RECEIVED = "RECEIVED"
    SESSION_CREATED = "SESSION_CREATED"
    RUNNING = "RUNNING"
    PR_OPENED = "PR_OPENED"
    CI_FAILED = "CI_FAILED"
    READY_FOR_REVIEW = "READY_FOR_REVIEW"
    MERGED = "MERGED"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
    ESCALATED = "ESCALATED"


ACTIVE_STATUSES = {
    TaskStatus.SESSION_CREATED,
    TaskStatus.RUNNING,
    TaskStatus.PR_OPENED,
    TaskStatus.CI_FAILED,
}

POLLABLE_STATUSES = ACTIVE_STATUSES

TERMINAL_STATUSES = {
    TaskStatus.MERGED,
    TaskStatus.COMPLETED,
    TaskStatus.FAILED,
    TaskStatus.ESCALATED,
}


class TriggerSource(str, enum.Enum):
    GITHUB_WEBHOOK = "github_webhook"
    MANUAL_API = "manual_api"
    SCAN = "scan"
    SCHEDULED = "scheduled"


class TaskKind(str, enum.Enum):
    REMEDIATION = "remediation"
    SMOKE_TEST = "smoke_test"


class RemediationTask(Base):
    __tablename__ = "remediation_tasks"
    __table_args__ = (
        UniqueConstraint("github_repository", "github_issue_number", name="uq_repo_issue"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)

    github_delivery_id: Mapped[str] = mapped_column(String(64), unique=True, index=True)
    github_repository: Mapped[str] = mapped_column(String(255))
    github_issue_number: Mapped[int] = mapped_column(Integer)
    github_issue_url: Mapped[str] = mapped_column(String(512))

    issue_title: Mapped[str] = mapped_column(String(512))
    issue_type: Mapped[str] = mapped_column(String(64), default="unknown")
    issue_labels: Mapped[str | None] = mapped_column(Text, nullable=True)
    task_kind: Mapped[str] = mapped_column(String(32), default=TaskKind.REMEDIATION.value)
    trigger_source: Mapped[str | None] = mapped_column(String(32), nullable=True)

    devin_session_id: Mapped[str | None] = mapped_column(String(128), nullable=True)
    devin_session_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    devin_status: Mapped[str | None] = mapped_column(String(32), nullable=True)
    devin_status_detail: Mapped[str | None] = mapped_column(String(64), nullable=True)
    devin_origin: Mapped[str | None] = mapped_column(String(32), nullable=True)
    devin_service_user_id: Mapped[str | None] = mapped_column(String(128), nullable=True)
    devin_tags: Mapped[str | None] = mapped_column(Text, nullable=True)

    status: Mapped[TaskStatus] = mapped_column(
        Enum(TaskStatus), default=TaskStatus.RECEIVED, index=True
    )

    pr_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    pr_state: Mapped[str | None] = mapped_column(String(64), nullable=True)

    retry_count: Mapped[int] = mapped_column(Integer, default=0)
    max_retries: Mapped[int] = mapped_column(Integer, default=3)

    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    merged_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    failure_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    completion_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    escalation_reason: Mapped[str | None] = mapped_column(Text, nullable=True)

    failure_type: Mapped[str | None] = mapped_column(String(32), nullable=True)
    ci_classification_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    ci_check_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    ci_check_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    ci_conclusion: Mapped[str | None] = mapped_column(String(64), nullable=True)
    ci_failure_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    ci_repair_attempts: Mapped[int] = mapped_column(Integer, default=0)
    last_ci_check_run_id: Mapped[int | None] = mapped_column(Integer, nullable=True)
    ci_repair_message_sent_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    ci_repair_verified_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    ci_passed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    ci_non_code_failure_count: Mapped[int] = mapped_column(Integer, default=0)

    acu_used: Mapped[float | None] = mapped_column(Float, nullable=True)
    acu_source: Mapped[str | None] = mapped_column(String(32), nullable=True)
    acu_verified: Mapped[bool] = mapped_column(Boolean, default=False)

    session_size: Mapped[str | None] = mapped_column(String(8), nullable=True)
    num_user_messages: Mapped[int | None] = mapped_column(Integer, nullable=True)
    num_devin_messages: Mapped[int | None] = mapped_column(Integer, nullable=True)
    insights_status: Mapped[str | None] = mapped_column(String(32), nullable=True)
    insights_json: Mapped[str | None] = mapped_column(Text, nullable=True)

    remediation_outcome: Mapped[str | None] = mapped_column(String(16), nullable=True)
    root_cause: Mapped[str | None] = mapped_column(Text, nullable=True)
    implementation_summary: Mapped[str | None] = mapped_column(Text, nullable=True)
    structured_result_json: Mapped[str | None] = mapped_column(Text, nullable=True)
    blocker: Mapped[str | None] = mapped_column(Text, nullable=True)
    playbook_id: Mapped[str | None] = mapped_column(String(128), nullable=True)

    merge_notification_sent: Mapped[bool] = mapped_column(default=False)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

File: frontend/src/types/task.ts
export type TaskStatus =
  | "RECEIVED"
  | "SESSION_CREATED"
  | "RUNNING"
  | "PR_OPENED"
  | "CI_FAILED"
  | "READY_FOR_REVIEW"
  | "MERGED"
  | "COMPLETED"
  | "FAILED"
  | "ESCALATED"

export type TestCategory =
  | "pre_fix_reproduction"
  | "post_fix_validation"
  | "regression_test"
  | "general_test"
  | "ci_validation"

export interface TestPerformed {
  command: string
  result: "passed" | "failed" | "skipped" | "not_run"
  category?: TestCategory | null
}

export interface StructuredResultPayload {
  tests_performed?: TestPerformed[]
  residual_risks?: string[]
  pr_url?: string | null
  raw?: Record<string, unknown>
}

export interface InsightsClassification {
  category: string | null
  confidence: number | null
  programming_languages: string[]
  tools_and_frameworks: string[]
}

export interface InsightsIssue {
  id: string
  title: string
  issue: string
  impact: string
  label: string
}

export interface InsightsTimelineEvent {
  title: string
  description: string
  color: string
  issue_id: string | null
}

export interface InsightsActionItem {
  action_item: string
  type: string
  issue_id: string | null
}

export interface Task {
  id: number
  github_delivery_id: string
  github_repository: string
  github_issue_number: number
  github_issue_url: string
  issue_title: string
  issue_type: string
  issue_labels: string | null
  task_kind: string
  trigger_source: string | null
  devin_session_id: string | null
  devin_session_url: string | null
  devin_status: string | null
  devin_status_detail: string | null
  devin_origin: string | null
  devin_service_user_id: string | null
  devin_tags: string | null
  status: TaskStatus
  pr_url: string | null
  pr_state: string | null
  retry_count: number
  max_retries: number
  started_at: string | null
  completed_at: string | null
  merged_at: string | null
  failure_reason: string | null
  completion_reason: string | null
  escalation_reason: string | null
  failure_type: string | null
  ci_classification_reason: string | null
  ci_check_name: string | null
  ci_check_url: string | null
  ci_conclusion: string | null
  ci_failure_at: string | null
  ci_repair_attempts: number
  max_ci_repair_attempts: number
  last_ci_check_run_id: number | null
  ci_repair_message_sent_at: string | null
  ci_repair_verified_at: string | null
  ci_passed_at: string | null
  ci_non_code_failure_count: number
  acu_used: number | null
  acu_source: string | null
  acu_verified: boolean
  session_size: string | null
  num_user_messages: number | null
  num_devin_messages: number | null
  insights_status: string | null
  insights_json: string | null
  remediation_outcome: string | null
  root_cause: string | null
  implementation_summary: string | null
  structured_result_json: string | null
  blocker: string | null
  playbook_id: string | null
  created_at: string
  updated_at: string
  mttr_seconds: number | null
}

export interface TaskListResponse {
  items: Task[]
  total: number
}
